Yelp/nrtsearch
# Add endpoint to clean up old index backups

# Change summary

Added endpoint `delete_index_backup` to clean up old nrtsearch indexes backups. The endpoint can be called on a regular basis to keep only backups from the last `n` days.  Changes include:
* Added 2 methods to `Archiver`:  `getVersionedResource` which gets all the version hashes for a particular resource, `deleteVersion` takes  a resource and a version hash and deletes that version
* Defined rpc for `delete_index_backup` endpoint
* Created `DeleteIndexBackupHandler` which decides what versions should be deleted basing on `nDays` parameter in the request


# Testing

## Manual testing procedure

Start (and restore) an index

```
curl -XPOST localhost:10003/v1/start_index -d '{
  "indexName" : "talk_v1",
   "mode": "PRIMARY",
   "restore" : {
     "serviceName" : "nrtsearch-generic",
     "resourceName": "talk_v1"
   }
}'
```

Create a few backups

```
curl -XPOST localhost:10003/v1/backup_index -d '{
	"indexName": "talk_v1", 
	"serviceName": "nrtsearch-generic-jblaszyk",
	"resourceName": "talk_v1"
}'
```

Call the new endpoint. Note: for the sake of **manual testing** I **changed days to minutes**. I was deleting versions older than 5 mins. 

```
curl -XPOST localhost:10003/v1/delete_index_backup -d '{
	"indexName": "talk_v1", 
	"serviceName": "nrtsearch-generic-jblaszyk", 
	"resourceName": "talk_v1", 
	"nDays": 5
}'
```

Manually verified on s3 that backups which should be deleted were deleted. 🎉

## Unit testing

Added unit tests for `Archiver` methods, and the `DeleteBackupIndexHandler`. 

Unit tests are ✅.
